### PR TITLE
Feature 5980 add logging debug callbacks to gfal

### PR DESCRIPTION
**Pull request by maatthias**

### CODE REVIEW

**Comment by bari12** [comment]
Is this well tested?
@mlassnig please have a look

File: lib/rucio/rse/protocols/gfal.py [hunk 1]
```diff
@@ -167,6 +167,12 @@ def path2pfn(self, path):
         else:
             return path
 
+    def event_callback(self, event):
+        self.logger(logging.DEBUG, "[%s] %s %s %s" % (event.timestamp, event.domain, event.stage, event.description))
+
+    def monitor_callback(self, src, dst, average, instant, transferred, elapsed):
+        self.logger(logging.DEBUG, "[%4d] %.2fMB (%.2fKB/s)\r" % (elapsed, transferred / 1048576, average / 1024))
```
Comment (dchristidis) [inline lib/rucio/rse/protocols/gfal.py]: I would suggest using MiB and KiB instead (IEC vs SI prefixes).

To be clear, I meant in the log message, not the calculations.

File: lib/rucio/rse/protocols/gfal.py [hunk 2]
```diff
@@ -176,15 +182,20 @@ def connect(self):
         """
         self.logger(logging.DEBUG, 'connecting to storage')
 
+        self.__ctx = gfal2.creat_context()  # pylint: disable=no-member
+        self.__ctx.set_opt_string_list("SRM PLUGIN", "TURL_PROTOCOLS", ["gsiftp", "rfio", "gsidcap", "dcap", "kdcap"])
```
Comment (mlassnig) [inline lib/rucio/rse/protocols/gfal.py]: this is missing https and davs

Comment (mlassnig) [reply lib/rucio/rse/protocols/gfal.py]: (I know, it was wrong even before :-)

Comment (vingar) [reply lib/rucio/rse/protocols/gfal.py]: https;gsiftp;root ? to be coherent with TURL_PROTOCOLS defined here https://fts3-atlas.cern.ch:8449/fts3/ftsmon/#/config/gfal2

File: lib/rucio/rse/protocols/gfal.py [hunk 3]
```diff
@@ -176,15 +182,20 @@ def connect(self):
...
+        self.__ctx.set_opt_string("XROOTD PLUGIN", "XRD.WANTPROT", "gsi,unix")
+        self.__ctx.set_opt_boolean("XROOTD PLUGIN", "NORMALIZE_PATH", False)
+
         if 'RUCIO_CLIENT_MODE' in os.environ:
             gfal2.set_verbose(gfal2.verbose_level.verbose)
+            params = self.__ctx.transfer_parameters()
+            params.event_callback = self.event_callback
+            params.monitor_callback = self.monitor_callback
+            params.overwrite = True
```
Comment (vingar) [inline lib/rucio/rse/protocols/gfal.py]: The overwrite option must be removed here.

File: lib/rucio/rse/protocols/gfal.py (hunk 2 above)
Comment (maatthias) [reply lib/rucio/rse/protocols/gfal.py]: So adding these protocols to the list of strings is enough?
```
self.__ctx.set_opt_string_list("SRM PLUGIN", "TURL_PROTOCOLS", ["gsiftp", "rfio", "gsidcap", "dcap", "kdcap", "https", "davs", "root"])
```

Comment (mlassnig) [reply lib/rucio/rse/protocols/gfal.py]: @mpatrascoiu is the order of protocols here important for gfal/davix?

Comment (maatthias) [reply lib/rucio/rse/protocols/gfal.py]: Is this conversation resolved?

**Comment by bari12** [comment]
Coming back to this, I think this is still useful, but as once mentioned, it would be good to separate this from normal debug mode, possibly over a rucio.cfg switch?

**Comment by bari12** [comment]
@maatthias I am closing this for now. I still think it is useful functionality, but it needs some changes.

**Pull request closed without merging**